llvm/utils/repo/Makefile

```diff
@@ -0,0 +1,121 @@
+#
+# Makefile for building LLVM libraries using musl-prepo libc (requires GNU make)
+# prerequisite: 1) installed the musl-prepo libc library in the $(MUSL) directory.
+#               2) built llvm-project-prepo in the $(LLVM_BUILD) directory. 
+#
+
+# The path of the directory in which llvm-project-prepo sources is checked out from git.
+LLVM_REPO = ${HOME}/github/llvm-project-prepo
+TOOLCHAIN_FILE=$(LLVM_REPO)/llvm/utils/repo/repo.cmake
+# The path of the llvm-project-prepo build directory.
+LLVM_BUILD = ${LLVM_REPO}/build
+# The path of the directory in which llvm libraries are installed.
+LLVM = ${HOME}/LLVM
+# The path of the directory in which musl libc is installed.
+MUSL = ${HOME}/musl
+
+libunwind.a:
+	if [ -d "$(LLVM_REPO)/build_libunwind" ];then                         \
+		rm -rf $(LLVM_REPO)/build_libunwind;                          \
+	fi
+	cd $(LLVM_REPO)  &&                                                   \
+	mkdir build_libunwind && cd build_libunwind  &&                       \
+	cmake -D musl=Yes                                                     \

```

Comment:
Running cmake from within a makefile is still… odd. Normally, we show how to run cmake and build on the project’s home page (e.g. [here (for pstore)](https://github.com/SNSystems/pstore#building) or [here (for llvm-project-prepo)](https://github.com/SNSystems/llvm-project-prepo#building-the-compiler or [your work on musl-prepo](https://github.com/MaggieYingYi/musl-prepo) which substibutes `./configure` for cmake).

Is it really necessary to run cmake four times (once for each target)? I would imagine that a single run could produce a build suitable for then building any of the libraries.

Reply:
> 
> 
> Running cmake from within a makefile is still… odd. Normally, we show how to run cmake and build on the project’s home page (e.g. [here (for pstore)](https://github.com/SNSystems/pstore#building) or [here (for llvm-project-prepo)](https://github.com/SNSystems/llvm-project-prepo#building-the-compiler or [your work on musl-prepo](https://github.com/MaggieYingYi/musl-prepo) which substibutes `./configure` for cmake).
> 

Do you mean we could only provide a Wiki page in the llvm-project-prepo to explain how to build LLVM runtime libraries using musl libc library?

Reply:
> 
> 
>
> Is it really necessary to run cmake four times (once for each target)? I would imagine that a single run could produce a build suitable for then building any of the libraries.
>

There are several reasons for separating the target:

1) we cannot do cmake in a single run when using the musl libc library.

I tried the following command:

```
cmake -D musl=Yes  -D libcxx=Yes  -D CMAKE_BUILD_TYPE=Release  -D LLVM_PATH=../ -D LLVM_CONFIG_PATH=../build/bin/llvm-config -D LIBUNWIND_ENABLE_SHARED=No -D LIBUNWIND_ENABLE_STATIC=Yes  -D COMPILER_RT_DEFAULT_TARGET_ONLY=ON -D COMPILER_RT_DEFAULT_TARGET_ONLY=ON -D COMPILER_RT_BUILD_BUILTINS=ON -D COMPILER_RT_BUILD_SANITIZERS=OFF -D COMPILER_RT_BUILD_XRAY=OFF -D COMPILER_RT_BUILD_LIBFUZZER=OFF -D COMPILER_RT_BUILD_PROFILE=OFF -D libcxxabi_include=-I../build/lib/clang/11.0.0/include/ -D LIBCXXABI_ENABLE_EXCEPTIONS=OFF -D LIBCXXABI_USE_COMPILER_RT=YES  -D LIBCXXABI_ENABLE_SHARED=No -D LIBCXXABI_ENABLE_STATIC=Yes -D LIBCXX_ENABLE_SHARED=No -D LIBCXX_ENABLE_STATIC=Yes -D LIBCXX_CXX_ABI=libcxxabi -D LIBCXX_CXX_ABI_INCLUDE_PATHS=../libcxxabi/include -D LIBCXX_HAS_MUSL_LIBC=ON -D LLVM_ENABLE_LIBCXX=ON -D CMAKE_TOOLCHAIN_FILE=../llvm/utils/repo/repo.cmake -D CMAKE_INSTALL_PREFIX=~/LLVM -DLLVM_ENABLE_PROJECTS="libunwind;compiler-rt;libcxx;libcxxabi;" ../llvm
```

I got the following cmake failure. 
```
CMake Error at cmake/modules/CheckAtomic.cmake:55 (message):
  Host compiler must support std::atomic!
```

2) compiler-rt contains an implementation of an atomics library. By building targets separately, the issue can be avoid.

3) the libunwind.a is  a dependency of the C++ ABI library, and sometimes is a dependency of other runtimes.

4) the libcxx.a is also dependent on the libcxxabi.a when using the musl libc.

The detailed information could be found here: https://clang.llvm.org/docs/Toolchain.html.

Reply:
I think that this probably needs further refinement, but your solution to these issues seems to do the trick for the moment.

Reply:
> Do you mean we could only provide a Wiki page in the llvm-project-prepo to explain how to build LLVM runtime libraries using musl libc library?

Yes. We’ll obviously need to hook it up to CI soon as well.